modules/packages/ConcurrentMap.chpl

```diff
@@ -0,0 +1,1046 @@
+              var idx = (newBuckets.hash(k) % newBuckets.size:uint):int;
+              if newBuckets.buckets[idx].read() == nil {
+                newBuckets.buckets[idx].write(new unmanaged Bucket(newBuckets));
+              }
+              var buck = newBuckets.buckets[idx].read() : unmanaged Bucket(keyType, valType)?;
+              buck!.count += 1;
+              buck!.keys[buck!.count] = k;
+              buck!.values[buck!.count] = v;
+            }
+
+            // TODO: Need to pass this to 'EpochManager.deferDelete'
+            next!.lock.write(GARBAGE);
+            tok.deferDelete(next);
+            curr.buckets[idx].write(newBuckets: unmanaged Base(keyType, valType));
+            curr = newBuckets;
+            idx = (curr._hash(defaultHash) % (curr.buckets.size):uint):int;
+          }
+        }
+
+        if shouldYield then chpl_task_yield(); // If lock could not be acquired
+        shouldYield = true;
+      }
+      return retNil;
+    }
+
+    // TODO: RAII based Locks.
+    // Current iterator can be locked indefinitely if function breaks
+    // Eg: for i in map do break;
+    /*
+      Serially iterates over the key-value pairs of this map.
+
+      :yields: A tuple whose elements are a copy of one of the key-value
+               pairs contained in this map.
+    */
+    iter these() : (keyType, valType) {
+      var tok = getToken();
+      tok.pin();
+      // Data stored on recursionStack: PointerList, start, i
+      var recursionStack = new LockFreeStack(stackType);
+      var deferred : unmanaged DeferredNode(deferredType)?;
+      var restore = true;
+      var curr : unmanaged Buckets(keyType, valType)? = root;
+      var start = ((iterRNG.getNext())%(curr!.buckets.size):uint):int;
+      var startIndex = 0;
+
+      while (true) {
+        restore = true;
+        for i in startIndex..(curr!.buckets.size-1) {
+          var idx = (start + i)%curr!.buckets.size;
+          var bucketBase = curr!.buckets[idx].read();
+          if (bucketBase != nil) {
+            if (bucketBase!.lock.read() == E_AVAIL && bucketBase!.lock.compareAndSwap(E_AVAIL, E_LOCK)) {
+              var bucket = bucketBase : unmanaged Bucket(keyType, valType)?;
+              for j in 0..#bucket!.count do yield (bucket!.keys[j], bucket!.values[j]);
+              bucket!.lock.write(E_AVAIL);
+            } else if (bucketBase!.lock.read() == P_INNER) {
+              var stackElem = (curr, start, i);
+              recursionStack.push(stackElem);
+              curr = bucketBase : unmanaged Buckets(keyType, valType)?;
+              start = ((iterRNG.getNext())%(curr!.buckets.size):uint):int;
+              startIndex = 0;
+              restore = false;
+              break;
+            } else {
+              var deferredElem = (curr, idx);
+              var deferredNode = new unmanaged DeferredNode(deferredElem);
+              deferredNode.next = deferred;
+              deferred!.prev = deferredNode;
```

Comment:
The test `testEquality` hits a `!` halt on this line. I think this line needs to check if deferred is `nil` before updating it. After I tried doing this I started looping forever on the `m1 != m2` line in the test.